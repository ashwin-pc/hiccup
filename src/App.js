import React from 'react';
import { FeaturedSection } from './components/FeaturedSection'
import { CategorySection } from './components/CategorySection'
import { SearchProvider } from './components/SearchContext'
import { ConfigProvider } from './components/ConfigContext'
import { SearchBar } from './components/SearchBar'
import { ConfigEditor } from './components/ConfigEditor'

const App = ({ config }) => {
  
  return (
    <ConfigProvider config={config}>
      <SearchProvider>
        <SearchBar />
        <FeaturedSection />
        <CategorySection />
        <ConfigEditor />
      </SearchProvider>
    </ConfigProvider>
  )
}

export default App;
