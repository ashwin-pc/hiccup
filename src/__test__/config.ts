const config = {
    featured: [{
        name: 'Featured Link',
        background: '/assets/logo.png',
        link: 'http://google.com'
    }, {
        name: 'Another Feaured link',
        link: 'http://google.com',
        tags: 'space separated tags'
    }],
    categories: [{
        title: 'Category 1',
        links: [{
            name: 'Link 1',
            link: 'http://google.com'
        }, {
            name: 'Test Link 2',
            link: 'http://google.com'
        }]
    }, {
        title: 'Category 2',
        links: [{
            name: 'Link 1',
            link: 'http://google.com',
            tags: 'more searchable tags'
        }, {
            name: 'Link 2',
            link: 'http://google.com'
        }]
    }]
}

export {
    config,
    config as default,
}