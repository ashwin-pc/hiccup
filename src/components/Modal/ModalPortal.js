import React from 'react'
import ReactDOM from 'react-dom'

const modalRoot = document.getElementById('modal-root');

class ModalPortal extends React.Component {
    render() {
        return ReactDOM.createPortal(
            this.props.children,
            modalRoot,
        );
    }
}

export {
    ModalPortal,
    ModalPortal as default
}