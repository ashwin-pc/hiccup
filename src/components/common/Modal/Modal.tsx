import React, {
  useState,
  useCallback,
  useEffect,
  FunctionComponent,
  MouseEventHandler,
} from 'react'
import { ReactComponent as CloseIcon } from './times-solid.svg'
import styles from './index.module.css'

export interface ModalProps {
  show?: boolean
  className?: string
  onClose?: Function
}

const Modal: FunctionComponent<ModalProps> = ({
  show = false,
  children,
  className,
  onClose,
}) => {
  const [open, setOpen] = useState(show)
  const handleClose = useCallback(() => {
    setOpen(false)
    onClose && onClose()
  }, [onClose])

  useEffect(() => {
    setOpen(show)
  }, [show])

  return open ? (
    <div className={styles.container}>
      <ModalBackdrop onClick={handleClose} />
      <ModalPopup className={className} onClose={handleClose}>
        {children}
      </ModalPopup>
    </div>
  ) : null
}

interface IModalBackdrop {
  onClick: MouseEventHandler
}

const ModalBackdrop: FunctionComponent<IModalBackdrop> = ({ onClick }) => (
  <div className={styles.backdrop} onClick={onClick} />
)

interface IModalPopup {
  onClose: MouseEventHandler
  className?: string
}

const ModalPopup: FunctionComponent<IModalPopup> = ({
  className,
  children,
  onClose,
}) => (
  <div className={[styles.modal, className || styles.default].join(' ')}>
    {children}
    <button className={styles.close} onClick={onClose}>
      <CloseIcon />
    </button>
  </div>
)

export { Modal, Modal as default }
