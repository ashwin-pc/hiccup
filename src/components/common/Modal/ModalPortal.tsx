import React from 'react'
import ReactDOM from 'react-dom'

const MODAL_ROOT_ID = 'modal-root'
class ModalPortal extends React.Component {
  render() {
    const modalRoot = document.getElementById(MODAL_ROOT_ID) as HTMLElement

    return ReactDOM.createPortal(this.props.children, modalRoot)
  }
}

export { ModalPortal, MODAL_ROOT_ID, ModalPortal as default }
