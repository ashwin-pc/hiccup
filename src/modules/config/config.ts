import { remoteService } from './remotes'
import { ConfigEntity, ConfigParams } from './types'
import validate from './validate'

/**
 * Config
 * A config is a remote or local entity that contains data of the links and categories
 */
export class Config {
  public readonly id: string
  public title: string
  public data?: ConfigEntity
  public readonly readonly: boolean
  public remote: ConfigParams['remote']
  public edited: boolean
  public error?: string

  constructor(params: ConfigParams) {
    const { id, title, remote, readonly, data, edited, error } = params

    if (!error || data) {
      // Validate the data
      const [valid, errorMessage, path] = validate(data)
      if (!valid) {
        throw new Error(
          `Invalid config: ${errorMessage} at ${path}\nConfig: ${JSON.stringify(
            data,
            null,
            2
          )}`
        )
      }
    }

    this.id = id
    this.title = title
    this.remote = remote
    this.readonly =
      readonly || remoteService.getRemote(remote.type).readonly || false
    this.data = data
    this.edited = edited || false
    this.error = error
  }

  get params(): ConfigParams {
    return {
      id: this.id,
      title: this.title,
      remote: this.remote,
      readonly: this.readonly,
      data: this.data,
      edited: this.edited,
      error: this.error,
    }
  }
}
