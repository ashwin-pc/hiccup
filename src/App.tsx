import { ConfigEntity } from 'modules/config/Config'
import { FeaturedSection } from './components/FeaturedSection'
import { CategorySection } from './components/CategorySection'
import { SearchProvider } from './components/SearchContext'
import { ConfigProvider } from './components/ConfigContext'
import { SearchBar } from './components/SearchBar'
import { ConfigEditor } from './components/ConfigEditor'
import { Hotkeys } from './components/Hotkeys'
import { ConnectedEditLinkModal } from 'components/EditLinkModal'

export const ROOT_ID = 'root'

interface Props {
  config?: ConfigEntity
}

const App = ({ config }: Props) => (
  <ConfigProvider config={config}>
    <SearchProvider>
      <SearchBar />
      <FeaturedSection />
      <CategorySection />
      <ConnectedEditLinkModal />
      <ConfigEditor />
      <Hotkeys />
    </SearchProvider>
  </ConfigProvider>
)

export default App
