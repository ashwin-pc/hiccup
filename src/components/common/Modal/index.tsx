import { FunctionComponent } from 'react'
import { ModalPortal } from './ModalPortal'
import { Modal, ModalProps } from './Modal'
import styles from './index.module.css'

const ConnectedModal: FunctionComponent<ModalProps> = (props) => (
  <ModalPortal>
    <Modal {...props} />
  </ModalPortal>
)

export {
  ConnectedModal as Modal,
  ConnectedModal,
  styles,
  ConnectedModal as default,
}
