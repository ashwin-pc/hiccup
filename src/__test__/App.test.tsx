import React from 'react'
import { render, screen } from '@testing-library/react'
import { config } from './config'
import App, { ROOT_ID } from '../App'
import { MODAL_ROOT_ID } from 'components/common/Modal/ModalPortal'

// TODO: Not a a very effective test, will need to replace this soon. leaving it in place until the other tests are setup
describe('Test hiccup', () => {
  test('should render page correctly', async () => {
    const rootEle = document.body.appendChild(createElement(ROOT_ID))
    document.body.appendChild(createElement(MODAL_ROOT_ID))

    const { container } = render(<App config={config} />, {
      container: rootEle,
    })

    await screen.findByText('Featured Link')
    expect(container).toMatchSnapshot()
  })

  function createElement(id: string) {
    const ele = document.createElement('div')
    ele.id = id
    return ele
  }
})
