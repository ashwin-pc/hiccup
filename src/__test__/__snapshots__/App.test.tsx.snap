// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test hiccup should render page correctly 1`] = `
<div
  id="root"
>
  <div
    class="search-container"
  >
    <input
      autocomplete="off"
      class="search"
      name="search"
      placeholder="Search   ...... or use Shift + Tab for URL bar"
      type="text"
      value=""
    />
    <svg
      class="icon search-icon"
      style="width: 10px;"
    >
      search-solid.svg
    </svg>
    <div
      class="results"
    />
  </div>
  <div
    class="section featured"
  >
    <a
      class="link"
      href="http://google.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <li
        class="card container "
        style="background-image: url(./assets/logo.png);"
      >
        Featured Link
      </li>
    </a>
    <a
      class="link"
      href="http://google.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <li
        class="card container "
        style="background-image: url(./assets/card.png);"
      >
        Another Feaured link
      </li>
    </a>
  </div>
  <div
    class="section categories"
  >
    <div
      class="category"
    >
      <h1
        class="title"
      >
        Category 1
      </h1>
      <ul>
        <a
          class="link"
          href="http://google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <li
            class="card container "
          >
            <span
              class="name"
            >
              Link 1
            </span>
            <span
              class="link"
            >
              http://google.com
            </span>
          </li>
        </a>
        <a
          class="link"
          href="http://google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <li
            class="card container "
          >
            <span
              class="name"
            >
              Test Link 2
            </span>
            <span
              class="link"
            >
              http://google.com
            </span>
          </li>
        </a>
      </ul>
    </div>
    <div
      class="category"
    >
      <h1
        class="title"
      >
        Category 2
      </h1>
      <ul>
        <a
          class="link"
          href="http://google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <li
            class="card container "
          >
            <span
              class="name"
            >
              Link 1
            </span>
            <span
              class="link"
            >
              http://google.com
            </span>
          </li>
        </a>
        <a
          class="link"
          href="http://google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <li
            class="card container "
          >
            <span
              class="name"
            >
              Link 2
            </span>
            <span
              class="link"
            >
              http://google.com
            </span>
          </li>
        </a>
      </ul>
    </div>
  </div>
  <div
    class="config-actions-container"
  >
    <button
      class="button"
    >
      <svg
        class="icon icon"
        style="width: 20px;"
      >
        pen-solid.svg
      </svg>
    </button>
    <button
      class="button"
    >
      <svg
        class="icon icon"
        style="width: 20px;"
      >
        cog-solid.svg
      </svg>
    </button>
  </div>
</div>
`;
